pkg/install/stack/kubermatic-master/migrations.go

```diff
@@ -0,0 +1,110 @@
+// The keys are updated on the master cluster and are then synced downstream by the project
+// controller. This syncing will fail as long as the seed clusters are not also updated (at least
+// the CRDs), so in this sense it's safe to do it on the master.
+// This function can be removed in KKP 2.22.
+func (*MasterStack) migrateUserSSHKeyProjects(ctx context.Context, client ctrlruntimeclient.Client, logger logrus.FieldLogger, opt stack.DeployOptions) error {
+	keys := &kubermaticv1.UserSSHKeyList{}
+	if err := client.List(ctx, keys); err != nil {
+		return fmt.Errorf("failed to list UserSSHKeys: %w", err)
+	}
+
+	apiVersion := kubermaticv1.SchemeGroupVersion.String()
+	kind := kubermaticv1.ProjectKindName
+
+	// determine if there are keys that need to be migrated
+	needMigration := false
+	for _, key := range keys.Items {
+		if key.Spec.Project == "" {
+			needMigration = true
+			break
+		}
+	}
+
+	// An old webhook could reject any changes to fields that it doesn't understand,
+	// so we must remove the problematic webhooks prior to the migration; to ensure
+	// the KKP operator doesn't bring them back, we have to scale it down first.
+	// Once the migration is over and the regular installation/upgrade has finished,
+	// the new operator will deploy a new webhook for us.
+	if needMigration {
+		logger.Info("Temporarily removing UserSSHKey webhook…")
+
+		if err := util.ShutdownDeployment(ctx, logger, client, KubermaticOperatorNamespace, KubermaticOperatorDeploymentName); err != nil {
+			return fmt.Errorf("failed to temporarily shut down the KKP operator: %w", err)
+		}
+
+		if err := util.WaitForAllPodsToBeGone(ctx, logger, client, KubermaticOperatorNamespace, KubermaticOperatorDeploymentName, 3*time.Minute); err != nil {
+			return fmt.Errorf("failed to wait for KKP operator to shut down: %w", err)
+		}
+
+		if err := util.RemoveMutatingWebhook(ctx, logger, client, common.UserSSHKeyAdmissionWebhookName); err != nil {
+			return fmt.Errorf("failed to remove the mutating webhook for UserSSHKeys: %w", err)
+		}
+	}
+
+	// Now that nobody should be blocking update operations on SSH keys, we can continue
+	// with the migration.
+	for _, key := range keys.Items {
+		if key.Spec.Project != "" {
+			continue
+		}
+
+		projectID := ""
+		for _, ref := range key.OwnerReferences {
+			if ref.APIVersion == apiVersion && ref.Kind == kind {
+				if projectID != "" {
+					return fmt.Errorf("key %s has multiple owner references to Projects, this should not be possible; reduce the owner refs to a single Project reference", key.Name)
+				}
+
+				projectID = ref.Name
+			}
+		}
+
+		if projectID == "" {
+			return fmt.Errorf("key %s no project owner reference, cannot determine project association", key.Name)
+		}
+
+		oldKey := key.DeepCopy()
+		key.Spec.Project = projectID
+
+		if err := client.Patch(ctx, &key, ctrlruntimeclient.MergeFrom(oldKey)); err != nil {
+			return fmt.Errorf("failed to update key %s: %w", key.Name, err)
+		}
+	}
```

Comment:
Not that it matters too much, but IMO this should happen within the `if needMigration {}` block. No key should be migrated if no migration is needed, but it seems unnecessary to loop over the keys a second time if that isn't needed.